File: python/xviz_avs/builder/metadata.py
import logging
from typing import Union

import numpy as np
from easydict import EasyDict as edict
from xviz_avs.builder.base_builder import (CATEGORY, CATEGORY_T,
                                           COORDINATE_TYPES,
                                           COORDINATE_TYPES_T, SCALAR_TYPE,
                                           SCALAR_TYPE_T, build_object_style,
                                           build_stream_style)
from xviz_avs.builder.ui_builder import XVIZUIBuilder
from xviz_avs.message import XVIZMessage
from xviz_avs.v2.session_pb2 import LogInfo, Metadata, StreamMetadata


class XVIZMetadataBuilder:
    """
    # Reference
    [@xviz/builder/xviz-metadata-builder]/(https://github.com/uber/xviz/blob/master/modules/builder/src/builders/xviz-metadata-builder.js)
    """
    def __init__(self):
        self._logger = logging.getLogger("xviz")
        self._data = Metadata(version="2.0.0")
        self._temp_ui_builder = None
        self._reset()

    def get_data(self) -> Metadata:
        self._flush()

        metadata = self._data

        if self._temp_ui_builder:
            panels = self._temp_ui_builder.get_ui()

            for panel_key in panels.keys():
                metadata.ui_config[panel_key].name = panel_key
                metadata.ui_config[panel_key].config.update(panels[panel_key])

        return metadata

    def get_message(self) -> XVIZMessage:
        return XVIZMessage(metadata=self.get_data())

    def start_time(self, time: float) -> 'XVIZMetadataBuilder':
        self._data.log_info.start_time = time
        return self

    def end_time(self, time: float) -> 'XVIZMetadataBuilder':
        self._data.log_info.end_time = time
        return self

    def ui(self, ui_builder: XVIZUIBuilder) -> 'XVIZMetadataBuilder':
        self._temp_ui_builder = ui_builder
        return self

    def stream(self, stream_id: str) -> 'XVIZMetadataBuilder':
        if self._stream_id:
            self._flush()

        self._stream_id = stream_id
        return self

    def category(self, category: Union[CATEGORY_T, str]):
        '''
        Assign category for the stream. Used for validation in XVIZBuilder and not required for data.
        '''
        if isinstance(category, int):
            self._temp_stream.category = category
        elif isinstance(category, str):
            self._temp_stream.category = CATEGORY.Value(category.upper())
        else:
            self._logger.error("Invalid value type for category!")
        return self

    def type(self, t: Union[SCALAR_TYPE_T, str]):
        '''
        Assign primitive type for the stream. Used for validation in XVIZBuilder and not required for data.
        '''
        if isinstance(t, int):
            self._temp_type = t
        elif isinstance(t, str):
            self._temp_type = SCALAR_TYPE.Value(t.upper())
        else:
            self._logger.error("Invalid value type for category!")
        return self

    def source(self, source: str):
        self._temp_stream.source = source
        return self

    def unit(self, u: str):
        self._temp_stream.units = u
        return self

    def coordinate(self, coordinate: COORDINATE_TYPES_T):
        self._temp_stream.coordinate = coordinate
        return self

    def transform_matrix(self, matrix: Union[list, np.ndarray]):
        matrix = np.array(matrix).ravel()
        self._temp_stream.transform.extend(matrix.tolist())
        return self

    def pose(self, position={}, orientation={}):
        raise NotImplementedError() # TODO: implement transformation

    def stream_style(self, style: dict):
        self._temp_stream.stream_style.MergeFrom(build_stream_style(style))
        return self

    def style_class(self, name: str, style: dict):
        if not self._stream_id:
            self._logger.error('A stream must set before adding a style rule.')
            return self
        style_class = self._temp_stream.style_classes.add()
        style_class.name = name
        style_class.style.MergeFrom(build_object_style(style))
        return self

    def log_info(self, data: dict):
        self._data.log_info.MergeFrom(LogInfo(**data))
        return self

    def _flush(self):
        if self._stream_id:
            stream_data = self._temp_stream

            if stream_data.category in [1, 5]:
                stream_data.primitive_type = self._temp_type
            elif stream_data.category in [2, 3]:
                stream_data.scalar_type = self._temp_type

            self._data.streams[self._stream_id].MergeFrom(stream_data)

        self._reset()

    def _reset(self):
        self._stream_id = None
        self._temp_stream = StreamMetadata()
        self._temp_type = None

File: cpp/src/builder/metadata.cpp
/*
 * File: metadata.cpp
 * Author: Minjun Xu (mjxu96@gmail.com)
 * File Created: Tuesday, 24th December 2019 8:15:24 pm
 */

#include "xviz/builder/metadata.h"

using namespace xviz;

XVIZMetadataBuilder::XVIZMetadataBuilder() {
  data_ = std::make_shared<Metadata>();
  // TODO UI BUILDER
  data_->set_version("2.0.0");
  Reset();
}

std::shared_ptr<Metadata> XVIZMetadataBuilder::GetData() {
  Flush();

  if (ui_ != nullptr) {
    auto panels = ui_->GetUI();
    auto ui_config = data_->mutable_ui_config();
    for (const auto& [panel_key, ui] : panels) {
      xviz::v2::UIPanelInfo panel;
      panel.set_name(panel_key);
      google::protobuf::Struct s;
      google::protobuf::util::JsonStringToMessage(
          ui.dump(), &s, google::protobuf::util::JsonParseOptions());
      panel.mutable_config()->MergeFrom(std::move(s));
      ui_config->insert({panel_key, panel});
    }
  }

  return data_;
}

XVIZMessage XVIZMetadataBuilder::GetMessage() { return XVIZMessage(GetData()); }

XVIZMetadataBuilder& XVIZMetadataBuilder::Stream(const std::string& stream_id) {
  if (!stream_id_.empty()) {
    Flush();
  }
  stream_id_ = stream_id;
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::StartTime(double time) {
  data_->mutable_log_info()->set_start_time(time);
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::EndTime(double time) {
  data_->mutable_log_info()->set_end_time(time);
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::UI(const XVIZUIBuilder& ui_builder) {
  return UI(std::make_shared<XVIZUIBuilder>(ui_builder));
}

XVIZMetadataBuilder& XVIZMetadataBuilder::UI(XVIZUIBuilder&& ui_builder) {
  return UI(std::make_shared<XVIZUIBuilder>(std::move(ui_builder)));
}

XVIZMetadataBuilder& XVIZMetadataBuilder::UI(
    const std::shared_ptr<XVIZUIBuilder>& ui_builder_ptr) {
  ui_ = ui_builder_ptr;
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::Source(const std::string& source) {
  temp_stream_.set_source(source);
  return *this;
}

// XVIZMetadataBuilder& XVIZMetadataBuilder::Source(std::string&& source) {
//   temp_stream_.set_source(std::move(source));
//   return *this;
// }

// XVIZMetadataBuilder& XVIZMetadataBuilder::Source(const char* source) {
//   temp_stream_.set_source(source);
//   return *this;
// }

XVIZMetadataBuilder& XVIZMetadataBuilder::Unit(const std::string& unit) {
  temp_stream_.set_units(unit);
  return *this;
}

// XVIZMetadataBuilder& XVIZMetadataBuilder::Unit(std::string&& unit) {
//   temp_stream_.set_units(std::move(unit));
//   return *this;
// }

// XVIZMetadataBuilder& XVIZMetadataBuilder::Unit(const char* unit) {
//   temp_stream_.set_units(unit);
//   return *this;
// }

XVIZMetadataBuilder& XVIZMetadataBuilder::Category(xviz::Category category) {
  temp_stream_.set_category(category);
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::Coordinate(
    CoordinateType coordinate_type) {
  temp_stream_.set_coordinate(coordinate_type);
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::Type(Primitive primitive_type) {
  type_ = (int)primitive_type;
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::Type(ScalarType scalar_type) {
  type_ = (int)scalar_type;
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::TransformMatrix(
    const std::vector<double>& matrix) {
  if (matrix.size() != 16u) {
    XVIZ_LOG_ERROR("Transform matrix should be a 4x4 matrix");
    return *this;
  }
  for (auto v : matrix) {
    auto v_ptr = temp_stream_.mutable_transform()->Add();
    *v_ptr = v;
  }
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::StreamStyle(
    const std::string& style_str) {
  auto stream_style = temp_stream_.mutable_stream_style();
  auto style_stream_ptr = JsonStringToStyleStream(style_str);
  if (xviz::StreamMetadata::PrimitiveType_IsValid(type_)) {
    ValidateStyle((xviz::StreamMetadata::PrimitiveType)type_, style_stream_ptr);
    temp_stream_.mutable_stream_style()->MergeFrom(*style_stream_ptr);
  } else {
    XVIZ_LOG_WARNING(
        "Before calling StreamStyle(), you must set the primitive type for "
        "this stream %s",
        stream_id_.c_str());
  }
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::StyleClass(
    const std::string& name, const std::string& style_str) {
  return StyleClass(name, nlohmann::json::parse(style_str));
}

XVIZMetadataBuilder& XVIZMetadataBuilder::StyleClass(
    const std::string& name, const nlohmann::json& style_json) {
  auto new_style_class = temp_stream_.add_style_classes();
  new_style_class->set_name(name);
  auto style_object_ptr = xviz::JsonObjectToStyleObject(style_json);
  new_style_class->mutable_style()->MergeFrom(*style_object_ptr);
  return *this;
}

XVIZMetadataBuilder& XVIZMetadataBuilder::LogInfo(double start_time,
                                                  double end_time) {
  data_->mutable_log_info()->set_start_time(start_time);
  data_->mutable_log_info()->set_end_time(end_time);
  return *this;
}

void XVIZMetadataBuilder::Reset() {
  stream_id_ = "";
  temp_stream_ = StreamMetadata();
  type_ = -1;
}

void XVIZMetadataBuilder::Flush() {
  if (stream_id_.empty()) {
    Reset();
    return;
  }

  auto category = temp_stream_.category();
  if (category == xviz::StreamMetadata::PRIMITIVE ||
      category == xviz::StreamMetadata::FUTURE_INSTANCE) {
    if (type_ == -1) {
      XVIZ_LOG_WARNING(
          "Did not set type for category: %s in stream: %s, available types "
          "are: %s",
          xviz::StreamMetadata::Category_Name(category).c_str(),
          stream_id_.c_str(),
          xviz::AllEnumOptionNames(
              xviz::StreamMetadata::PrimitiveType_descriptor())
              .c_str());
    } else {
      if (!xviz::StreamMetadata::PrimitiveType_IsValid(type_)) {
        XVIZ_LOG_ERROR("Type %s in category %d is invalid.",
                       xviz::StreamMetadata::Category_Name(category).c_str(),
                       type_);
        return;
      } else {
        temp_stream_.set_primitive_type((Primitive)(type_));
      }
    }
  } else if (category == xviz::StreamMetadata::TIME_SERIES ||
             category == xviz::StreamMetadata::VARIABLE) {
    if (type_ == -1) {
      XVIZ_LOG_WARNING(
          "Did not set type for category: %s in stream: %s, avaiable types "
          "are: %s",
          xviz::StreamMetadata::Category_Name(category).c_str(),
          stream_id_.c_str(),
          xviz::AllEnumOptionNames(
              xviz::StreamMetadata::ScalarType_descriptor())
              .c_str());
    } else {
      if (!xviz::StreamMetadata::ScalarType_IsValid(type_)) {
        XVIZ_LOG_ERROR("Type %s in category %d is invalid.",
                       xviz::StreamMetadata::Category_Name(category).c_str(),
                       type_);
        return;
      } else {
        temp_stream_.set_scalar_type((ScalarType)type_);
      }
    }
  }

  auto streams = data_->mutable_streams();
  (*streams)[stream_id_].MergeFrom(temp_stream_);

  Reset();
}